### ISSUE TITLE
Sandboxed genrule breaks after 0.3.2 upgrade

### ISSUE THREAD

**Issue by ekuefler** [bazelbuild/bazel#1921]
I have a fairly simple genrule that worked in 0.3.1 and fails in 0.3.2:

```
genrule(
    name = "build-properties",
    outs = ["build.properties"],
    cmd = "echo commitNumber=$(commitNumber) > $@",
)
```

It fails with the following error:

```
ERROR: /master/src/main/java/com/ekuefler/BUILD:19:1: Executing genrule //src/main/java/com/ekuefler:build-properties failed: linux-sandbox failed: error executing command /root/.cache/bazel/_bazel_root/c9fadd9098376b2dddd3da4175fdad4b/execroot/master/_bin/linux-sandbox ... (remaining 5 argument(s) skipped)
src/main/tools/linux-sandbox-pid1.cc:380: "remount(NULL, /root/.cache/bazel/_bazel_root/c9fadd9098376b2dddd3da4175fdad4b/bazel-sandbox/9ef26712-52ba-44ad-91a8-9c8914705335-0/tmp/tmp/.X11-unix, NULL, 2101281, NULL)": No such file or directory
Target //src/main/java/com/ekuefler:build-properties failed to build
Use --verbose_failures to see the command lines of failed build steps.
```

It works if I pass `--genrule_strategy=standalone`. I'm running this build inside of (privileged) docker container.

**Comment by hermione521** [comment]
May I ask how you get $(commitNumber) since it's not in cmd..

**Comment by ekuefler** [comment]
It's passed via a `--define` flag in the build system, e.g. `bazel build src/main/java/com/ekuefler:build-properties --define commitNumber=123`.

I get the same error if I replace `$(commitNumber)` with `123` in the build file, so I don't think that's related.

**Comment by hermione521** [comment]
Sorry, I'm very new to docker. Can you provide an image or is it easy to create one by myself?

**Comment by ekuefler** [comment]
Okay, this is weirder than expected. I got it to reproduce in a minimal container, but only if I run that container with a volume mount to something in my `/tmp` directory (was the X11 directory in the example I posted, but it seems to work for anything under tmp).

It should be pretty easy for you to repro. [Install docker](https://docs.docker.com/engine/installation/), create a new directory, and create a file named `Dockerfile` in it with the following contents:

```
FROM debian:jessie

# Install Java
ENV JAVA_HOME=/opt/jdk PATH=/opt/jdk/bin:$PATH
RUN apt-get update \
    && apt-get install -y gcc nano wget zip \
    && wget \
        -qO- \
        --no-check-certificate \
        --no-cookies \
        --header "Cookie: oraclelicense=accept-securebackup-cookie" \
        http://download.oracle.com/otn-pub/java/jdk/8u45-b14/jdk-8u45-linux-x64.tar.gz | tar xvz -C /opt \
    && mv /opt/jdk1.8.0_45 /opt/jdk

# Install Bazel
ENV PATH $PATH:/opt/bazel/bin
RUN wget -O /tmp/bazel.sh https://github.com/bazelbuild/bazel/releases/download/0.3.2/bazel-0.3.2-installer-linux-x86_64.sh \
    && chmod +x /tmp/bazel.sh \
    && /tmp/bazel.sh --prefix=/opt/bazel --bazelrc=/root/.bazelrc \
    && rm /tmp/bazel.sh
```

In that directory, run `docker build .`. You should see a build process install the JDK and Bazel, then print out a line like `Successfully built a571f8fe01b9` at the very end to give you a container ID. Run it like this:

```
you@yourcomputer$ docker run -it --privileged -v /tmp/whatever:/tmp/whatever  a571f8fe01b9
root@0e0a37b91690:/# mkdir workspace
root@0e0a37b91690:/# cd workspace/
root@0e0a37b91690:/workspace# touch WORKSPACE
root@0e0a37b91690:/workspace# nano BUILD
root@0e0a37b91690:/workspace# bazel build :demo --verbose_failures --sandbox_debug
```

Note the `-v` flag that mounts something in `/tmp` - it doesn't matter what subdirectory you use, as long as it's in `/tmp`. In the BUILD file, paste this:

```
genrule(
    name = "demo",
    outs = ["demo.txt"],
    cmd = "echo hello > $@",
)
```

When I do this, I get this output:

```
root@0e0a37b91690:/workspace# bazel build :demo --verbose_failures --sandbox_debug
INFO: Found 1 target...
ERROR: /workspace/BUILD:1:1: Executing genrule //:demo failed: linux-sandbox failed: error executing command 
  (cd /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/execroot/workspace && \
  exec env - \
    PATH=/opt/jdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/bazel/bin \
  /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/execroot/workspace/_bin/linux-sandbox @/root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/linux-sandbox.params -- /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; echo hello > bazel-out/local-fastbuild/genfiles/demo.txt').
src/main/tools/linux-sandbox.cc:183: linux-sandbox-pid1 has PID 964
src/main/tools/linux-sandbox-pid1.cc:241: tmpfs: /tmp
src/main/tools/linux-sandbox-pid1.cc:251: working dir: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/execroot/workspace
src/main/tools/linux-sandbox-pid1.cc:181: CreateTarget(root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/execroot/workspace, true)
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/proc
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/dev
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/dev/pts
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/dev/mqueue
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/dev/shm
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/dev/console
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/systemd
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/net_cls,net_prio
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/cpu,cpuacct
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/pids
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/cpuset
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/memory
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/blkio
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/perf_event
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/hugetlb
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/freezer
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/sys/fs/cgroup/devices
src/main/tools/linux-sandbox-pid1.cc:369: remount ro: /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/tmp/whatever
src/main/tools/linux-sandbox-pid1.cc:380: "remount(NULL, /root/.cache/bazel/_bazel_root/eab0d61a99b6696edb3d2aff87b585e8/bazel-sandbox/bfc3798f-fea7-4823-9c06-190a66bea0d7-0/tmp/tmp/whatever, NULL, 2101281, NULL)": No such file or directory
src/main/tools/linux-sandbox.cc:223: child exited normally with exitcode 1
Target //:demo failed to build
```

For some reason, the sandbox tries to mount the volume that was mounted in docker, and for some reason it fails. It works fine if I mount other directories outside of tmp. I see it's mounting into a directory that's also named tmp, so maybe something about having `tmp/tmp` is confusing it?

**Comment by ekuefler** [comment]
Still an issue on 0.4.0 and is preventing my upgrade, any idea what might be going wrong?

**Comment by philwo** [comment]
This was indeed a bug in the sandbox - it mounts a tmpfs on /tmp, which hides all mount points underneath /tmp that existed before, but then still tries to remount them read-only.

This was supposedly fixed by https://bazel-review.googlesource.com/#/c/6970/, which came after Bazel 0.4.0 was released. It is included from 0.4.1 onwards, so I think this is fixed now.

If it still doesn't work, please reopen.
